juju/juju
# Reject charms with unstorable data

Previously, charm data was accepted that generated field names which
couldn't be stored correctly in MongoDB. MongoDB doesn't check these in
*some* insert scenerios but will reject them in later updates and
queries.

Fixing this properly by escaping and unescaping everywhere is too
dangerous given the imminent 2.1 release, so problematic charm data 
is now rejected outright. It is unlikely that current charms will contain
metadata with invalid field names.

This PR adds some utility functions to the mongo/utils package. 
1. IsValidFieldName returns whether a field name is valid for use in a
MongoDB document.
2. CheckStorable checks if a document - and its subdocuments - uses valid
field names throughout.

Drive-by: apiserver/application.StoreCharmArchive was not returning the error when a the charm update to state failed.

### QA

Deployed some charms and confirmed they still work. Then deployed a charm with invalid charm data to confirm that it is rejected.

```
$ juju deploy ubuntu
Located charm "cs:ubuntu-10".
Deploying charm "cs:ubuntu-10".
menno@maiwa ~/canonical/repository/xenial 
$ juju deploy postgresql
Located charm "cs:postgresql-114".
Deploying charm "cs:postgresql-114".
menno@maiwa ~/canonical/repository/xenial 
$ juju deploy ./ubuntubad        
ERROR cannot upload charm: invalid charm data: "$foo" is not a valid field name
```
